### ISSUE TITLE
Check input types

### ISSUE THREAD

**Issue by OliviaLynn** [LSSTDESC/tables_io#79]
We should add explicit type checks to inputs and throw exceptions if they are the wrong type, with an explanation of what’s wrong.

An example of this in GalSim: [link](https://github.com/GalSim-developers/GalSim/blob/3a9ab46b3c11333cc5def004cbce3c74f6f186cf/galsim/gsobject.py#L920C9-L920C9).

This has come up recently with a relatively new user building a pipeline and receiving somewhat cryptic errors. I remember goldenspike being broken last October because some package it used had switched to jax arrays over the previous numpy arrays, so it looks like this may continue being a problem.

**Comment by rmandelb** [comment]
Thanks for opening this issue, Olivia!  Just to add to this (as I'm part of the project that inspired this issue):

In my experience, users outside of the development team may not always be aware of the right usage for given functions.  Providing the wrong type for the inputs is a really common failure mode.  If you explicitly check and raise a useful error about what's wrong, they can quickly learn the right usage.  If you do not do a type check, they often will get a cryptic error due to the code trying to do something that isn't allowed for their particular input somewhere within the function. While Olivia shared a specific example in GalSim, it was something we tried to do systematically, and it has been quite helpful for users.

I think the issue is relevant not just to `tables_io` (where it happens to have showed up) but also to the RAIL ecosystem, as DESC members will be using these codes for research projects meant to prepare for cosmological analyses using photo-z, and these checks will support those users' adoption and effective usage of the code.

**Comment by eacharles** [comment]
It would be good to clarify the scope of what you have in mind.   Are you more concerned with container type (I.e., Jax v numpy) or the data type (Ie float v double)? the first is relative straightforward, modolu the fact that python is duck-typed, so sometime “type checking” is not actually a well defined thing, the later is much more involved.On Jul 26, 2023, at 9:45 AM, Olivia R. Lynn ***@***.***> wrote:﻿
We should add explicit type checks to inputs and throw exceptions if they are the wrong type, with an explanation of what’s wrong.
An example of this in GalSim: link.
This has come up recently with a relatively new user building a pipeline and receiving somewhat cryptic errors. I remember goldenspike being broken last October because some package it used had switched to jax arrays over the previous numpy arrays, so it looks like this may continue being a problem.

**Comment by rmandelb** [comment]
Obviously I cannot speak for Olivia, but personally I was referring primarily to the case where the inputs are a particular class with methods that the function is going to try to use (like a pandas dataframe versus a qp.Ensemble versus a numpy array versus a list/tuple), rather than e.g. float vs. int.  The manifestation of the issue is the cryptic error that you get when the function tries to use some method of the class it's expecting, and it's preventable through an explicit check of the type.

**Comment by eacharles** [comment]
It’s a bit tricky, b/c in some cases we would very like to be able to handle a panda data frame OR and dictionary of numpy arrays.    That is what I was referring to about duck-typing.     
  

-e

**Comment by rmandelb** [comment]
Yes, completely understood that in some cases multiple types would work (we've dealt with that situation in GalSim before as well).  I won't be at the meeting next week but I'm happy to answer further questions here asynchronously.

**Comment by eacharles** [comment]
Do we have an example of these cryptic errors that people are receiving, and the data files that generated them.   I think that could help the discussion.

Tables_io is already trying to strike a balance between checking that things really are tables, and being able to support things that effectively act as tables, so it would be really helpful to have additional examples where this isn't working.

**Comment by eacharles** [comment]
The point about some downstream code trying to use some function doesn't exist is particulary tricky here.

The idea tables_io is that it makes it easy to read things that are "table-like" in the sense that provide certain functionality, like have len() and __getitem__ do preditable things.   We haven't been explict about what "table-like" means, and we could clarify that.  

On the other hand, some "table-like" classes provide additional functionality like join() or whatever, and in that case the code is going to crash when that function is called on a different type of table that doesn't support that function.  

I'm not sure the later case actually something we should be trying to catch in tables_io, as nice as it would be to be able to do that.  We should discuss this more.

**Comment by eacharles** [comment]
![Screen Shot 2023-07-25 at 10 58 39 AM](https://github.com/LSSTDESC/tables_io/assets/14844706/f0ee398c-de1c-43e1-b047-ffb59db7660b)

**Comment by eacharles** [comment]
That's the screenshot Olivia sent me relating to this issue.  
If I understand correctly, the issue is that `datafile`, which looks be some sort of DataHandle wrapping a qp.Ensemble object got passed to a ColumnMapper stage, which expects a catalog-like table, probably in 

It doesn't look like tables_io is involved in this.   

The docstring for ColumnMapper says:

```
    Utility stage that remaps the names of columns.

    Notes
    -----
    1. This operates on pandas dataframs in parquet files.

    2. In short, this does:
    `output_data = input_data.rename(columns=self.config.columns, inplace=self.config.inplace)`

```

    
However, the interactive routine, `__call__(self, data):` doesn't do explicit type-checking, (though it could, since it clearly expects a pandas.DataFrame object, and has an overly vague docstring, that says it tables table-like data, when actually it only takes DataFrames.

Anyway, if this was the source of the somewhat cryptic error messages that triggered this issue, I'm pretty sure the place we want fix is ColumnMapper in  rail_base .

**Comment by eacharles** [comment]
Can we confirm that this issue relates to the screen shot above?   If so, I think we should close it, and instead consider putting checks in the DataHandle classes in rail.
